src/net/neighbors.rs

```diff
@@ -0,0 +1,2557 @@
+        assert!(self.state == NeighborWalkState::GetHandshakesFinish);
+
+        // see if we got any replies 
+        let mut new_unresolved_handshakes = HashMap::new();
+        for (naddr, rh) in self.unresolved_handshake_neighbors.drain() {
+            let res = rh.try_recv();
+            let rh_naddr = naddr.clone();       // used below
+            let new_rh = match res {
+                Ok(message) => {
+                    match message.payload {
+                        StacksMessageType::HandshakeAccept(ref data) => {
+                            // success! do we know about this peer already?
+                            let neighbor_from_handshake = Neighbor::from_handshake(tx, message.preamble.peer_version, message.preamble.network_id, &data.handshake)?;
+                            let mut neighbor_opt = Neighbor::from_neighbor_address(tx, message.preamble.peer_version, message.preamble.network_id, block_height, &naddr)?;
+                            match neighbor_opt {
+                                Some(neighbor) => {
+                                    test_debug!("{:?}: already know about {:?}", &local_peer, &neighbor.addr);
+
+                                    // knew about this neighbor already
+                                    self.resolved_handshake_neighbors.insert(naddr, neighbor.clone());
+
+                                    // update our frontier as well
+                                    self.new_frontier.insert(neighbor.addr.clone(), neighbor);
+                                    neighbor_from_handshake.save_update(tx)?;
+                                },
+                                None => {
+                                    test_debug!("{:?}: new neighbor {:?}", &local_peer, &neighbor_from_handshake.addr);
+
+                                    // didn't know about this neighbor yet. Try to add it.
+                                    let added = neighbor_from_handshake.save(tx)?;
+                                    if !added {
+                                        // no more room in the db.  See if we can add it by
+                                        // evicting an existing neighbor once we're done with this
+                                        // walk.
+                                        let replaced_neighbor_slot_opt = NeighborWalk::find_replaced_neighbor_slot(tx, &neighbor_from_handshake.addr)?;
+
+                                        match replaced_neighbor_slot_opt {
+                                            Some(slot) => {
+                                                // if this peer isn't whitelisted, then consider
+                                                // replacing
+                                                if neighbor_from_handshake.whitelisted > 0 && (neighbor_from_handshake.whitelisted as u64) < get_epoch_time_secs() {
```

Comment:
Why don't we attempt to add the peer to the neighbor database if it's whitelisted? (Also, for clarity's sake, I'd recommend  making a `is_whitelisted()` function)